src/main/java/com/palantir/gradle/versions/GradleWorkarounds.java

```diff
@@ -124,5 +130,82 @@ static boolean isFailOnVersionConflict(Configuration conf) {
         return conflictResolution == org.gradle.api.internal.artifacts.configurations.ConflictResolution.strict;
     }
 
+    @SuppressWarnings("CyclomaticComplexity")
+    public static void makeEvaluationDependOnSubprojectsToBeEvaluated(Project rootProject) {
+        if (!rootProject.getGradle().getStartParameter().isConfigureOnDemand()
+                || rootProject.getGradle().getStartParameter().isWriteDependencyLocks()
+                // If Gradle is run from somewhere other than the root, the task location gets trickier to translate
+                // into the projects to use; this could be implemented in the future if there's demand
+                || !rootProject.getGradle().getStartParameter().getCurrentDir().equals(rootProject.getRootDir())) {
+            // Just configure every project
+            rootProject.getSubprojects().forEach(subproject -> rootProject.evaluationDependsOn(subproject.getPath()));
+            return;
+        }
+
+        Set<String> projectPathsToEval = new LinkedHashSet<>();
+        for (String taskPath : rootProject.getGradle().getStartParameter().getTaskNames()) {
+            if (!taskPath.contains(":")) {
+                // This is a task to be run in each project that defines it, e.g. "build". This should cause every
+                // project to be defined in a configuration-on-demand build.
```

Comment:
Am I right in thinking we need to do this because Gradle does not automatically do it for us when in configure on demand mode? I would have thought the configuration-on-demand feature in gradle should have the behaviour of evaluating every project if you run a general task like `./gradle build`, or it just plain wouldn't work?

Reply:
Or is this just ensuring the *ordering* of project evaluation is in the correct fashion for GCV to function?

Reply:
Gradle will configure those projects automatically, but not early enough in the build process for GCV's needs. Before this PR, we forced all the subprojects to be configured early with this line before proceeding:

https://github.com/palantir/gradle-consistent-versions/pull/657/files#diff-8f6332e4bf665c9817bd219407e5081faf2bd8e9b94d16b1e11d14bb2205f4e2L258

Configuring every subproject unconditionally, as that line does, would make CoD useless; but if we just remove that line with no replacement, some projects aren't configured before the point in time where we need them to know their dependencies. So we replace it with something that configures the right set of projects, or as close as we can get.

(This could perhaps work differently if writing locks happened during a task rather than during configuration, but I'm not going to try to make that large of a change, which I'm sure would cause other problems.)

Reply:
Ah, I see, makes sense - thanks for the explanation.